1460: Reservation buttons overhaul Round 2

holt83 opened the pull request:
#### Link to issue

https://platform.dandigbib.org/issues/1460

Continuation of the work in https://github.com/ding2/ding2/pull/1139

#### Description

This PR contains changes requested in previous PR and some small corrections:

1. Optimization of reservability checks. Now checks multiple materials at once - both in frontend and backend.
2. Do not check reservability for periodical reserve buttons to avoid making unnecessary requests.
3. Handle FBS errors when checking reservability to prevent fatal PHP error. 

#### Screenshot of the result

If your change affects the user interface you should include a screenshot of the result with the pull request.

#### Checklist

- [x] My complies with [our coding guidelines](../docs/code_guidelines.md).
- [x] My code passes our static analysis suite. If not then I have added a comment explaining why this change should be exempt from the code standards and process.
- [x] My code passes our continuous integration process. If not then I have added a comment explaining why this change should be exempt from the code standards and process.

#### Additional comments or questions

If you have any further comments or questions for the reviewer them please add them here.

kasperg commented on modules/ding_entity/ding_entity.api.php:
```diff
@@ -62,6 +62,55 @@ function hook_ding_entity_buttons($type, $entity, $view_mode, $widget) {
+ *   doesn't have a say for a particular entity and/or pseudo-class it must set
+ *   NULL to return something for that entity and signal that it is neutral.
+ */
+function hook_ding_entity_entities_is(array $entities, $class) {
```
I understand what you are trying to do here but I am not sure I agree about the approach.

Bundling multiple checks into a single request is nice and efficient but it makes it very hard for developers to know whether they should 1) use 2) implement `ding_entity_entities_is()` or `ding_entity_is()` for their modules.

One approach to this would be to fall back to `ding_entity_is()` for modules and classes which does not support `ding_entity_entities_is()`. This quickly becomes very complex and it may not be possible at all.

Another approach could be to implement a new provider hook á la `fbs_reservation_is_reservable($provider_ids)` where the implementation for FBS would be about the same as in `fbs_ding_entity_entities_is()`.

A third approach might be to use the `reservable` info from the `availability_items` provider hook. If implemented correctly by the provider this could actually provide the correct information. Note that this creates a dependency from ding_reservation against an availability provider making the whole situation a bit more muddy.

Personally I am leaning towards a new provider hook.

I could also be OK with multiple calls to `->is('reservable')` although it may decrease performance.

Whatever you decide on please remember to clean up. #1139 also introduced changes to `connie` and elsewhere which may not be needed anymore.

holt83 replied:
Thanks for the good explanation 👍 I understand your concerns.

I think it's important to get the performance benefit from the bundling. After reading your feedback I also lean towards a new provider hook, so I will implement that instead.

Regarding the changes to connie from #1139: I added the entity_is implementation to ensure to get reservation buttons when testing with that module. I havn't really used this module a lot, so this might be wrong. Would you say that it's not needed?

holt83 replied:
Done

kasperg commented on modules/ding_reservation/js/ding_reservation_reservability.js:
```diff
@@ -8,19 +8,26 @@
-        var entityId = reserveButton.data("entity-id");
+      var entityIds = [];
+      var selector = '.ting-object .reserve-button.check-reservability';
+      $(selector, context).once('check-reservability', function() {
```
If this class is purely for JS reasons then please prefix `check-reservability` with `js-`. With that in mind I think we can also drop the `.reserve-button` class from the selector - perhaps even `.ting-object`.

holt83 replied:
Done

holt83 commented:
@kasperg I updated the PR based:

- Use provider function for checking reservability instead of ding_entity pseudo classes check.
- Add js prefix to check-reservability
- Updated/cleaned up Connie
- Remove unneeded stuff

GitCop commented:
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 2359f6ae8a567bbc3db2e46353aa4379438d8520
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 2359f6ae8a567bbc3db2e46353aa4379438d8520
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

cableman commented on modules/ding_reservation/ding_reservation.module:
```diff
@@ -67,7 +67,7 @@ function ding_reservation_menu() {
   );
 
-  $items['ding_reservation/%ding_entity/is_reservable'] = array(
+  $items['ding_reservation/%/is_reservable'] = array(
```
Found the following tabel (so IE and Edge may have issues):

Browser    |  Address bar   
------------|------------------------------
Chrome      |    32779        
Android      |    8192           
Firefox         |   >64k          
Safari            | >64k           
IE11                | 2047           
Edge 16          |2047

holt83 replied:
Interesting question :)

I have tried to do some calculations:

We have 31 chars in "ding_reservation//is_reservable" and let's say the domain part is 50 chars to be sure. Then we have 1.966 chars left. Each faust is 8 chars and we also need a comma so:

1.966 / 9 = 218,44

So we should be able to handle 218 fausts even in IE11 and Edge. 

This could be a problem on frontpage with a lot carousels. But a lot of them are loaded async at different timings, so there will not be one single request that checks reservability for alle materials on frontpage. Rather it will be split up in 2-3 requests. 
Also considering that the materials in the carousels are lazy loaded on demand themselves I don't think carousels will cause issues with the limitation.

Another case to consider is news/events with material lists. If an administrator adds over 218 materials to a list there could be an issue in IE/Edge, since these will all be checked in one single request.

I guess we could split the request up in max 200 fausts in the client. What do you think?

kasperg replied:
You cannot assume what the length of the local ids is. For example the length of local ids for Icelandic libraries using DDELibra Aleph is 20 chars. For other library systems it might be something else.

You could chunk requests dynamically in the client based on length of local ids and browser but in increases the complexity somewhat. Another approach would be to convert it to a POST request.

If you come up with a method which chucks requests using GET then also remember to set appropriate headers which prevent caching of the response.

holt83 replied:
I think I'll go with the POST request.

But why prevent caching if I was using GET? Reservability is not something that changes often, so I see no problem with the response being cached. Care to elaborate why you think this is necessary?

holt83 replied:
> I do not know what it is for Ereolen

Doesn't matter. FBS know nothing about these ids and only operate on fausts. So it doesn't make sense to look up these ids.

holt83 replied:
Fixed! Using post request instead.

kasperg replied:
> But why prevent caching if I was using GET? Reservability is not something that changes often, so I see no problem with the response being cached. Care to elaborate why you think this is necessary?

You are right. I was mixing up reservability and availability.

holt83 replied:
> As afar as I know Ereolen uses Publizon. Not FBS.

Ah ok. Wasn't aware of that.

kasperg commented on modules/ding_reservation/ding_reservation.module:
```diff
@@ -143,9 +152,9 @@ function ding_reservation_ding_entity_buttons($type, $entity, $view_mode, $widge
                 ),
-                'id' => 'reservation-' . $entity->id,
-                'data-entity-id' => $entity->id,
+                'data-local-id' => $entity->localId,
```
:+1:

kasperg commented on modules/ding_entity/ding_entity.api.php:
```diff
@@ -57,8 +57,15 @@ function hook_ding_entity_collection_view(&$object, $view_mode) {
+    if (ding_provider_implements('reservation', 'is_reservable')) {
+      $reservable = ding_provider_invoke('reservation', 'is_reservable', [$entity->localId]);
+      $reservable = reset($reservable);
+      if ($reservable) {
```
I know what your intention is here but for correctness it would probably be a good idea to replace the two lines with something that ensures that we are checking availability for the same id that we passed into the function:

```php
if (!empty($reservable[$entity->localId])) {
```

holt83 replied:
👍

holt83 replied:
Fixed!

kasperg commented on modules/ding_reservation/ding_reservation.module:
```diff
@@ -113,10 +113,19 @@ function ding_reservation_ding_entity_menu(&$items, $type, $path, $index) {
+  $result = [];
+
+  // The reservation provider decides reservability.
+  if (ding_provider_implements('reservation', 'is_reservable')) {
```
As mentioned previously I think we can remove this check and simplify the code.

holt83 replied:
I agree with you that it shouldn't be necessary.  

Just want to mention that ding_reservation uses this pattern in a lot of other places. Do you have any guidelines on when to use/not to use?

holt83 replied:
Edit: After looking at code I see that the implements-pattern is only used when the provider callback is of optional nature.

holt83 replied:
Fixed!

kasperg commented on modules/ding_reservation/ding_reservation.module:
```diff
@@ -164,10 +173,14 @@ function ding_reservation_ding_entity_buttons($type, $entity, $view_mode, $widge
+        if (ding_provider_implements('reservation', 'is_reservable')) {
+          $reservable = ding_provider_invoke('reservation', 'is_reservable', [$entity->localId]);
+          $reservable = reset($reservable);
+          if ($reservable) {
```
See comments elsewhere.

holt83 replied:
Fixed!

kasperg commented on modules/ting/ting.admin.inc:
```diff
@@ -136,7 +136,7 @@ function ting_admin_reservable_settings($form_state) {
     '#type' => 'fieldset',
     '#title' => t('Reservation settings'),
-
+    '#tree' => FALSE,
```
Is this needed?

holt83 replied:
Last time [you questioned why I removed it](https://github.com/ding2/ding2/pull/1139#discussion_r292157752). Then i told you it was because it's default, but that maybe it would be good to keep for clarity. No response, but no biggie I just added it again.

kasperg commented on themes/ddbasic/template.php:
```diff
@@ -829,15 +829,15 @@ function ddbasic_process_ting_object(&$vars) {
             );
           }
-          if ($vars['object']->online_url) {
+          if ($vars['object']->is('online')) {
```
:+1:

holt83 commented on modules/ding_provider/connie/connie.reservation.inc:
```diff
@@ -187,3 +187,13 @@ function connie_reservation_branch_name($branch_id) {
+  // Make everything reservable.
+  $return = array_fill(0, count($provider_ids), TRUE);
+  $return = array_combine($provider_ids, $return);
+  return $return;
```
Thanks for the tip 👍

holt83 replied:
Fixed!

holt83 commented on modules/ding_entity/ding_entity.api.php:
```diff
@@ -57,8 +57,15 @@ function hook_ding_entity_collection_view(&$object, $view_mode) {
-    return array(ding_provider_get_form('ding_reservation_reserve_form', new DingReservationReservableEntity($entity), TRUE));
+  // Adds resevation buttons is entity is resevable.
+  if ($type == 'ding_entity' && $entity->is('library_material')) {
+    if (ding_provider_implements('reservation', 'is_reservable')) {
```
Will remove

holt83 replied:
Fixed!

holt83 commented:
@kasperg PR is updated. I also rebased to get Scutinizer working.

holt83 commented:
We discovered an additional issue with referenced materials on ding_news content missing reservation button, which we decided to handle here also: https://platform.dandigbib.org/issues/4415#note-17 

The problem is that ting_reference is using custom AJAX-callback which doesn't account for attached scripts during the rendering process in the callback.

I have chosen to include the reservability checking script manually in this case. I think it makes sense that it's the modules doing something non-standard Drupal, that should take care that the correct scripts is added.

An alternative solution would be just include the reservablilty checking script on every page in ding_reservation module.

kasperg commented:
I agree with the your solution to the missing JavaScript library.
